Bluetooth: Mesh: Add support for provisioning remote devices

tsvehagen opened the pull request:
### Known limitations:
- Only AUTH_METHOD_NO_OOB is supported currently. I think this is good enough for now and support for OOB can be added in another PR.
- ~~Add a mesh_provisioner sample to samples/bluetooth~~ I think mesh_shell is good enough.
- ~~Add documentation~~
- Only PB-ADV is supported. I think this is also good enough for now and support for PB-GATT can be added in another PR

### Questions:
1. Start and PubKey messages should be sent right after each other and currently the sending of the PubKey is triggered from gen_prov_ack() when the ACK for the Start message is received. This feels like a weird placement.. any other ideas for handling this?

2. ~~When provisioning is complete, the specification states that a LINK_CLOSE message should be sent at least three times. Any ideas how to handle this in a nice way? I guess we should not reset the link on Complete but rather when all the close messages have been sent.~~ FIXED

3. ~~Might it be a good idea to rename bt_mesh.dev_key to bt_mesh.local_dev_key and bt_mesh.dev_keys to bt_mesh.remote_dev_keys?~~ FIXED

4. Currently there is no way to build provisioner support but not device support. It could be possible to keep BT_MESH_PROV option and then have a BT_MESH_PROV_DEVICE default to y and a BT_MESH_PROV_PROVISIONER default to n. BT_MESH_PROV would include the common stuff while the others would be specific things for device and provisioner. Is this necessary or should device support always be enabled when BT_MESH_PROV is enabled?

5. ~~Storing of AppKey and NetKey supports being written by a timer. I guess it is to minimize flash operations. This is currently not implemented for DevKey storage. Should this be done and if so should I try to modify and reuse the key_update struct for DevKeys as well?~~ FIXED

### Test it:

I have used the mesh shell together with a nrf52840_pca10056 board and a nrf52_adafruit_feather board for testing. Simply build one with the BT_MESH_PROVISIONER option enabled.

On device:
mesh init
mesh uuid 1234
mesh pb-adv on

On provisioner:
mesh init
mesh provision 0 1
mesh provision-adv 1234 0 2 0
mesh get-comp

tsvehagen committed:
Bluetooth: Mesh: Add support for provisioner role over PB-ADV

Make it possible to provision devices over advertising bearer (PB-ADV).
Many messages in the provisioning protocol are the same for provisioner
and device so much of the code could be reused by only changing when
they are expected to arrive.

This introduces to concept of local and remote device keys. The models
for cfg_cli and cfg_srv have been updated to reflect this concept. Both
the send and receive path in the transport layer have been updated to
support encrypting/decrypting with local and remote device keys.

When a node has been provisioned it is stored in bt_mesh_net.nodes. If
CONFIG_BT_SETTINGS is enabled, they are also saved to settings. If the
callback node_added in bt_mesh_prov has been set, it will be called for
every node that gets provisioned. This includes when they are retrieved
from settings.

The configuration CONFIG_BT_MESH_NODE_COUNT controls how many nodes that
can be provisioned.

Signed-off-by: Tobias Svehagen <tobias.svehagen@gmail.com>

tsvehagen committed:
Bluetooth: Mesh: Add callback for unprovisioned device beacon

Adds the unprovisioned_beacon callback to the bt_mesh_prov structure.

Signed-off-by: Tobias Svehagen <tobias.svehagen@gmail.com>

tsvehagen committed:
Bluetooth: Mesh: Add bt_mesh_provision_adv API

Add API for supporting provisioning of devices over PB-ADV.

Signed-off-by: Tobias Svehagen <tobias.svehagen@gmail.com>

tsvehagen committed:
Bluetooth: Mesh: Add support for provisioner over PB-ADV to shell

Adds the command provision-adv to the mesh shell.

Signed-off-by: Tobias Svehagen <tobias.svehagen@gmail.com>

trond-snekvik commented:
Some input: 

I think provision_remote might be a bad name for provisioning other devices with PB-adv. The SIG wants to reserve this keyword for devices that are somewhere else in the mesh network, instead of right next to the provisioner, so that in the future, we can tunnel provisioning through the mesh. Ideally, `bt_mesh_provision` should probably be `bt_mesh_provision_self()`, so that this new API could take its place, but I guess it's a bit late for that. Perhaps `bt_mesh_provision_other`, `_by_uuid` or `_node` could be a good compromise? 

IMO question 3 should be handled by changing the new `bt_mesh_dev_key` structure to something like `bt_mesh_node`. It should probably align with the `bt_mesh_provision_remote` API name. The dev key is just one part of the structure, which actually represents a whole node, so I found the current naming a bit confusing. I suppose this structure could also replace the two first input parameters of the config client API functions, to avoid leaking the netkey-to-addr binding.

+1 for Q5.

Works very well on my nrf52_pca10040 setup. Confirmed working with the Nordic Mesh SDK as well.

tsvehagen commented:
Good input!

> I think provision_remote might be a bad name for provisioning other devices with PB-adv. The SIG wants to reserve this keyword for devices that are somewhere else in the mesh network, instead of right next to the provisioner, so that in the future, we can tunnel provisioning through the mesh. Ideally, `bt_mesh_provision` should probably be `bt_mesh_provision_self()`, so that this new API could take its place, but I guess it's a bit late for that. Perhaps `bt_mesh_provision_other`, `_by_uuid` or `_node` could be a good compromise?

Totally agree on the `bt_mesh_provision_self()` part but like you say, it might be a bit too late for that change :/ I think I like the `_node` name best but any of the above is fine by me. I'll wait for some more feedback.

> IMO question 3 should be handled by changing the new `bt_mesh_dev_key` structure to something like `bt_mesh_node`. It should probably align with the `bt_mesh_provision_remote` API name. The dev key is just one part of the structure, which actually represents a whole node, so I found the current naming a bit confusing.

Yes, that name is a bit confusing and I like the bt_mesh_node name. Should probably change the setting DevKey to Node as well then. Do you think the `bt_mesh_dev_key_find` should be renamed to `bt_mesh_node_find` as well?

> I suppose this structure could also replace the two first input parameters of the config client API functions, to avoid leaking the netkey-to-addr binding.

Do you mean passing in a `struct bt_mesh_node` to functions like `bt_mesh_cfg_comp_data_get`?

> +1 for Q5.

I'll look into that then.

> Works very well on my nrf52_pca10040 setup. Confirmed working with the Nordic Mesh SDK as well.

Nice to hear!

jhedberg commented:
> Ideally, bt_mesh_provision should probably be bt_mesh_provision_self(), so that this new API could take its place, but I guess it's a bit late for that

@trond-snekvik I don’t think it’s too late. That API has mainly been for internal use (called by PB-ADV or PB-GATT) and only used directly by apps for demo purposes. So I wouldn’t be against renaming it. That said, the new API for provisioning other devices should probably still get its own suffix. Btw, IIRC the spec uses “device” for unprovisioned entities and “node” for provisioned ones. I.e the act of provisioning makes a device into a node. 

I haven’t thought too deeply about this, but wouldn’t it make sense to name the APIs based on provisioning bearer? So we’d have provision_gatt() provision_adv() and then the current provision() would become provision_self() or provision_ local().

trond-snekvik commented:
> Do you think the bt_mesh_dev_key_find should be renamed to bt_mesh_node_find as well?

IMO yes, this'll make more sense when the usage of the node structure expands in the future (to config client or the configuration database @jhedberg mentioned in #17454, for instance).

> Do you mean passing in a struct bt_mesh_node to functions like bt_mesh_cfg_comp_data_get?

Yes, the cfg_cli.h signatures could look like `int bt_mesh_cfg_comp_data_get(struct bt_mesh_node * node, u8_t page, u8_t *status, struct net_buf_simple *comp);`. 
Since the first two parameters of the cfg_cli functions always have to be used with the correct combination of netkey and address, the application will have to either store them as a pair after they come out of the provisioning_complete callback, or pick them out of the dev_key array with the `bt_mesh_dev_key_find()` function. Either way, the current cfg_cli API leaks this binding, increasing complexity in the user app and opening up for more bugs, and the bt_mesh_dev_key structure could alleviate this. It could also help make it less awkward to configure its own server.
This would change would break the API, and it's probably out of scope for this PR. I'm not confident that it's worth doing, but I thought it might be worth discussing :)

Related: The provisioning_complete callback is now serving two purposes: being provisioned and provisioning others. This should probably either be reflected in its documentation, or a separate callback should be made for the provisioner, preferably with the `struct bt_mesh_dev_key` as a parameter instead of the current netkey-address tuple. The need for the new `is_provisioned` flag in shell.c highlights this problem, I think.

> wouldn’t it make sense to name the APIs based on provisioning bearer?

I like this, it'll future proof the API for new bearers, and fits well with the spec.

trond-snekvik commented on include/bluetooth/mesh/main.h:
```diff
@@ -128,6 +128,15 @@ struct bt_mesh_prov {
+	 *
+	 * @param uuid UUID
+	 */
+	void        (*unprovisioned_beacon)(u8_t uuid[16]);
```
Could you expose the URI hash here as well?

tsvehagen replied:
Yes and probably the OOB Information as well?

trond-snekvik replied:
Oh yeah, good point!

tsvehagen replied:
Added OOB info and URI hash.

tsvehagen commented:
> > Do you think the bt_mesh_dev_key_find should be renamed to bt_mesh_node_find as well?
> 
> IMO yes, this'll make more sense when the usage of the node structure expands in the future (to config client or the configuration database @jhedberg mentioned in #17454, for instance).

Ok I'll go for that.

> > Do you mean passing in a struct bt_mesh_node to functions like bt_mesh_cfg_comp_data_get?
> 
> Yes, the cfg_cli.h signatures could look like `int bt_mesh_cfg_comp_data_get(struct bt_mesh_node * node, u8_t page, u8_t *status, struct net_buf_simple *comp);`.

Sounds like a plan but like you say, it probably belongs in another PR.

> Related: The provisioning_complete callback is now serving two purposes: being provisioned and provisioning others. This should probably either be reflected in its documentation, or a separate callback should be made for the provisioner, preferably with the `struct bt_mesh_dev_key` as a parameter instead of the current netkey-address tuple. The need for the new `is_provisioned` flag in shell.c highlights this problem, I think.

Yes that is the reason for the `is_provisioned` flag in the shell. Either adding another callback or a `bool local` flag to the existing one sounds like a good idea. Is it better with a new callback?

> > wouldn’t it make sense to name the APIs based on provisioning bearer?
> 
> I like this, it'll future proof the API for new bearers, and fits well with the spec.

Yea I also think that makes sense. Good idea @jhedberg.

trond-snekvik commented:
> Either adding another callback or a bool local flag to the existing one sounds like a good idea. Is it better with a new callback?

That's what I would do, but I'm still taking in _The Zephyr Way™️_ of doing things, so your judgement is probably better than mine 😄

tsvehagen commented:
> > Either adding another callback or a bool local flag to the existing one sounds like a good idea. Is it better with a new callback?
> 
> That's what I would do, but I'm still taking in _The Zephyr Waytm_ of doing things, so your judgement is probably better than mine

Haha I doubt that! Let's wait and see what @jhedberg has to say about it :)

jhedberg commented:
> > > Either adding another callback or a bool local flag to the existing one sounds like a good idea. Is it better with a new callback?
> > 
> > That's what I would do, but I'm still taking in _The Zephyr Waytm_ of doing things, so your judgement is probably better than mine
> Haha I doubt that! Let's wait and see what @jhedberg has to say about it :)

Definitely a new callback 😄 It could also be an ephemeral one given as a parameter to the `provision()` API rather than being registered separately.

tsvehagen commented:
> > > > Either adding another callback or a bool local flag to the existing one sounds like a good idea. Is it better with a new callback?
> > > 
> > > 
> > > That's what I would do, but I'm still taking in _The Zephyr Waytm_ of doing things, so your judgement is probably better than mine
> 
> > Haha I doubt that! Let's wait and see what @jhedberg has to say about it :)
> 
> Definitely a new callback  It could also be an ephemeral one given as a parameter to the `provision()` API rather than being registered separately.

I added a `complete_node` callback for now at least.

carlescufi commented:
@tsvehagen thanks for the PR. We are going to release 2.0 in a few days, and after that the merge window will be open again for merging PRs that are not bugfixes.
Do you have an estimate as to when you will be able to continue the work here? I am setting the milestone for 2.1 tentatively in the meantime.

tsvehagen commented:
> @tsvehagen thanks for the PR. We are going to release 2.0 in a few days, and after that the merge window will be open again for merging PRs that are not bugfixes.
> Do you have an estimate as to when you will be able to continue the work here? I am setting the milestone for 2.1 tentatively in the meantime.

I'm back from vacation and can kind of continue whenever. I think I need some more feedback from @jhedberg on what is needed to get it merged.

Milestone 2.1 sounds good to me.

jhedberg commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -1465,6 +1750,41 @@ void bt_mesh_pb_adv_recv(struct net_buf_simple *buf)
+	bt_rand(&link.id, sizeof(link.id));
+	link.tx.id = 0x7F;
+	link.node->net_idx = net_idx;
+	link.node->addr = addr;
```
Doesn't this get initialised by `bt_mesh_node_alloc()`? I'd just pass both `addr` and `net_idx`to `bt_mesh_node_alloc()` and initialize both there.

jhedberg commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -1465,6 +1750,41 @@ void bt_mesh_pb_adv_recv(struct net_buf_simple *buf)
+		return -ENOMEM;
+	}
+
+
```
Redundant empty line

jhedberg commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -88,6 +88,28 @@ int bt_mesh_provision(const u8_t net_key[16], u16_t net_idx,
+		return -EINVAL;
+	}
+
+	if (!bt_mesh_is_addr_free(addr)) {
```
How will this work if the remote device has multiple elements? Wouldn't it make sense to allocate the address only once you've received the Provisioning Capabilities? In such a case I suppose the `bt_mesh_provision_adv()` wouldn't take an `addr` parameter at all, rather you could report it back (along with number of elements) through the provisioning complete callback.

tsvehagen replied:
Yea having this check here doesn't really solve the whole problem since we don't yet know the number of elements. Agree that it should be removed.

The `addr` argument could be removed and instead use the first free address that fits. I don't really know if it is of interest to be able to specify an address but maybe it could be an ok solution to pass a `addr_hint` or something. If that is not free then choose the next one that is free.

trond-snekvik replied:
I think there's value in having an address parameter, but perhaps it could be possible to pass invalid address to make it allocate dynamically?

tsvehagen replied:
Yea I think there is value in being able to specify an address as well. What do you think @jhedberg? Maybe keep the `addr` parameter but if it is 0 (BT_MESH_ADDR_UNASSIGNED) we select the lowest possible address?

Also, in order to have an effective algorithm to find the lowest possible address I think the list of nodes would have to be stored in a sorted list. What do you guys think would be the best solution for finding a free address range?

jhedberg replied:
Sounds good regarding keeping the address and the 0 value. Would need to think about the sorting a bit more.

tsvehagen replied:
I have implemented support for this without having a sorted list but as mentioned in the comment it could be implemented more effectively if they were sorted. I don't think it's that big a deal though.

jhedberg commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -434,6 +443,25 @@ config BT_MESH_SHELL
 	  the console.
 
+config BT_MESH_NODES
+	bool "Support for saving node information"
```
Should this depend on `BT_MESH_PROVISIONER`? Is this about runtime or flash storage? If it's the latter I suppose a `BT_SETTINGS` dependency is also needed. Update: I see you're doing the reverse, i.e. selecting this option from `BT_MESH_PROVISIONER`. See my other comment about that.

I understand the need to come up with a short name for this feature, and "NODES" is not a bad attempt. I find it slightly confusing though since the local device is typically a node itself. To me at least, this all sounds like a provisioning database, so would perhaps `BT_MESH_PROV_DB` sound intuitive?

tsvehagen replied:
> I understand the need to come up with a short name for this feature, and "NODES" is not a bad attempt. I find it slightly confusing though since the local device is typically a node itself. To me at least, this all sounds like a provisioning database, so would perhaps `BT_MESH_PROV_DB` sound intuitive?

I like BT_MESH_PROV_DB. Do you think the name `node` or `nodes` in the code is ok? I can really think of a good name to distinguish between the local node and other nodes.

jhedberg replied:
`node`/`nodes` in the code is fine with me, however if there's no good justification for keeping a separate `BT_MESH_PROV_DB` Kconfig option around (i.e. a reason why you'd sometimes want one but not the other) I'd just stick to a single `BT_MESH_PROVISIONER` option.

tsvehagen replied:
`BT_MESH_NODES` is removed but `BT_MESH_NODE_COUNT` is still left. Do you think this is ok or should it be maybe `BT_MESH_PROV_DB_SIZE` or something?

jhedberg commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -32,6 +32,15 @@ config BT_MESH_PB_ADV
+	bool "Provisioner support"
+	select BT_ECC
+	select BT_MESH_PROV
+	select BT_MESH_NODES
```
Do you really need the separate `BT_MESH_NODES` option? In what kind of scenario would one want to enable `BT_MESH_NODES` but not enable `BT_MESH_PROVISIONER`?

tsvehagen replied:
It's not really needed no, it just felt like a natural split for some reason and also it is currently two separate commits. I can merge them and just leave the BT_MESH_PROVISIONER.

jhedberg commented on include/bluetooth/mesh/main.h:
```diff
@@ -157,6 +171,17 @@ struct bt_mesh_prov {
+	 *  @param net_idx NetKeyIndex given during provisioning.
+	 *  @param addr Primary element address.
+	 */
+	void        (*complete_node)(u16_t net_idx, u16_t addr);
```
Should there be an element count parameter here?

I'm also not completely happy with the name of the callback. Just brainstorming, but what would you think about `node_added`, `new_node` or `node_ready`?

tsvehagen replied:
> Should there be an element count parameter here?

Yes I think it should be included.

> I'm also not completely happy with the name of the callback. Just brainstorming, but what would you think about `node_added`, `new_node` or `node_ready`?

`node_added` sounds good. Could even be `prov_db_node_added` to be clear but maybe it's a bit too long.

tsvehagen replied:
Named it `node_added` and added the `num_elem` param,

jhedberg commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -208,10 +208,22 @@ static const struct bt_mesh_comp comp = {
+
+static void prov_complete_node(u16_t net_idx, u16_t addr)
+{
+
```
Redundant empty line

jhedberg commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -1108,7 +1344,28 @@ static void prov_data(const u8_t *data)
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		bt_mesh_store_node(node);
+	}
```
Empty line after this (always after `}` unless the next line is also a `}`)

jhedberg reviewed (changes requested):
Generally this looks very promising, thanks! It seems we may indeed be able to do everything necessary in a single `prov.c` file.

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1,4 +1,5 @@
-/** @file
+
+	   /** @file
```
Looks like something went wrong here

tsvehagen replied:
He sure does

tsvehagen commented:
I think this is ready for merge or at least ready to review for a merge.

There is a problem with bluetooth on master which seems to make a device receive its own packets which messes up the transactions. To avoid this problem please build the `mesh_shell` with `CONFIG_LOG_IMMEDIATE=n`.

@jhedberg @trond-snekvik, could you please have a look and see what you think again :)

jhedberg commented:
> I think this is ready for merge or at least ready to review for a merge.

Thanks, I'll try to find time this week for a proper review.

> To avoid this problem please build the mesh_shell with CONFIG_LOG_IMMEDIATE=n.

Wouldn't #19512 fix that?

tsvehagen commented:
> > To avoid this problem please build the mesh_shell with CONFIG_LOG_IMMEDIATE=n.
> 
> Wouldn't #19512 fix that?

Aah, yes. I rebased on master again and `LOG_IMMEDIATE=n` by default now.

jhedberg commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -656,13 +671,45 @@ static int sdu_recv(struct bt_mesh_net_rx *rx, u32_t seq, u8_t hdr,
+			return 0;
+		}
+
+#if defined(CONFIG_BT_MESH_PROVISIONER)
```
It should be possible to put this inside a `if (IS_ENABLED(CONFIG_BT_MESH_PROVISIONER)) {` block and thereby avoid the #ifdef.

jhedberg reviewed (changes requested):
This is starting to look quite close to something mergeable. Thanks a lot for the work so far!

The comments I had was mainly refactoring the node database handling into its own c-file, and taking more advantage of `IS_ENABLED()` instead of #ifdefs (check today's backlog of the #bluetooth slack channel for some discussion regarding this, if you're interested)

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -971,20 +1107,101 @@ static void prov_input_complete(const u8_t *data)
+	prov_buf_init(&pdu, PROV_DATA);
+
+	err = bt_mesh_prov_encrypt(session_key, nonce, data,
+				   net_buf_simple_add(&pdu, 33));
```
Couldn't the encryption be done in place? You could save 25 bytes of stack memory and use net_buf_simple for populating the fields.

tsvehagen replied:
Ehm I'm not really sure... I guess it depends on how the `bt_mesh_ccm_encrypt` functions. Do you know if it's ok to have `msg` and `out_msg` point to the same buffer?

trond-snekvik replied:
That should be fine, both app_encrypt and net_encrypt does it.

tsvehagen replied:
Done

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -971,20 +1107,101 @@ static void prov_input_complete(const u8_t *data)
-static void prov_random(const u8_t *data)
+static void send_random(void)
 {
 	PROV_BUF(rnd, 16);
```
This buffer is one byte too small to fit the random data plus packet type. Looks like an existing bug.

tsvehagen replied:
Fixed, good catch :)

trond-snekvik replied:
This has been addressed in #19789, btw.

trond-snekvik commented on subsys/bluetooth/mesh/prov.c:
```diff
@@ -108,12 +109,16 @@
-	REMOTE_PUB_KEY,        /* Remote key has been received */
+	WAIT_PUB_KEY,          /* Waiting for local PubKey to be generated */
 	LINK_ACTIVE,           /* Link has been opened */
+	LINK_ACK_RECVD,        /* Ack for link has been received */
```
The link ack is only checked to avoid duplicate callback calls, it's never required for operation. It should probably be treated like the LINK_ACTIVE flag to ensure that we wait for the link ack before we accept any other behavior. 

It'd probably be easiest to change it to `WAIT_LINK_ACK`, and postpone `LINK_ACTIVE` until after the ack is received.

tsvehagen replied:
The only "problem" is that LINK_ACTIVE is used with `atomic_test_and_set_bit`. If for example the link is already waiting for an ack, ie `WAIT_LINK_ACK` is set, it shouldn't be possible to start another provisioning. That means that any call to start provisioning also has to check against the `WAIT_LINK_ACK` and you then get into a race if it is split between two atomic calls.

One way could be to use the `atomic_cas` maybe. Like in `bt_mesh_pb_gatt_open` for example, `atomic_cas(link.flags, 0, LINK_ACTIVE)` and in `bt_mesh_pb_adv_open`, `atomic_cas(link.flags, 0, PROVISIONER | WAIT_LINK_ACK)`. This would ensure that it is not waiting to ack.

Another option would be to not use `LINK_ACTIVE` as the flag for checking if provisioning is ongoing or not but instead introduce another flag for that.

Any ideas @trond-snekvik?

trond-snekvik replied:
I hadn't thought about that, that's a good point. Sounds like your last suggestion would be the cleanest. Something `LINK_OPEN` to replace `LINK_ACK_RECVD`, but used in both provisioner and provisionee, for instance?

I guess you could also check if either `WAIT_LINK_ACK` or `LINK_ACTIVE` is set when allocating a provisioning instance?

tsvehagen replied:
> I hadn't thought about that, that's a good point. Sounds like your last suggestion would be the cleanest. Something `LINK_OPEN` to replace `LINK_ACK_RECVD`, but used in both provisioner and provisionee, for instance?

So that would kind of mean renaming `LINK_ACTIVE` to `LINK_OPEN` and then renaming `LINK_ACK_RECVD` to `LINK_ACTIVE`? :thinking: 

> I guess you could also check if either `WAIT_LINK_ACK` or `LINK_ACTIVE` is set when allocating a provisioning instance?

There's not really a way to test for two flags so that could also introduce a race. I think we could use `atomic_cas()` though since the flags should be 0 when starting out.

trond-snekvik commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -116,6 +116,20 @@ struct mod_pub_val {
+static struct node_update {
+	u16_t addr;
+	bool clear;
+} node_updates[CONFIG_BT_MESH_NODE_COUNT];
```
We'd save a lot of memory by using two bitfields instead, one indicating change, and one indicating clear/store.

tsvehagen replied:
How do you mean? Similar to the `struct key_update`? I'm not sure if a change flag is really needed for nodes?

trond-snekvik replied:
I meant changing the design from dynamically allocated list of pending changes to bitfields where each bit represents an index in the node array. We'd need one bit to represent "needs storing" and one to represent the type of change, as this can't be inferred from the `bt_mesh_node` alone, which might get reallocated before we're able to do the change. The settings entries must also be stored by index, instead of by address.

This'll reduce RAM requirements for each node by 15%, and reduce the amount of searching in this module. It'll also create a tighter coupling between the `bt_mesh` struct and the settings module though, so it's not all good :smile:

tsvehagen replied:
Okay I see what you mean. I think it maybe could be another PR?

trond-snekvik replied:
Absolutely! :slightly_smiling_face:

tsvehagen replied:
Yea that's true but I think it might be okay since this is a new addition and I think the API etc. will most likely change anyway. What do you think? @trond-snekvik

trond-snekvik commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -684,6 +698,58 @@ static int vnd_mod_set(const char *name, size_t len_rd,
+
+		node = bt_mesh_node_find(addr);
+		if (node) {
+			bt_mesh_node_del(node, false);
```
Is there a scenario where there's a need to do cleanup on load?

tsvehagen replied:
Please elaborate :)

trond-snekvik replied:
Isn't that what's happening here? I'm asking whether it's actually necessary.

tsvehagen replied:
Oh I see :p

Well I guess there could be a need if a node has been removed. Currently there is no way to remove a node but I think it makes sense to add that possibility. I'm just talking about removing the node from the list, not doing the whole network key changing thingy.

If I understand the settings correctly if a node has been added and then removed, `node_set()` will be called once for the first entry, which will allocate a new node. Then `node_set()` will be called again but with len 0 meaning it has been removed.

jhedberg replied:
Now that #19541 has been merged, I don't think you will get history replay anymore. So in new settings handlers I don't think we need to handle that case (we should also gradually start removing unnecessary code from the existing handlers)

jhedberg commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -734,6 +741,150 @@ u32_t bt_mesh_next_seq(void)
+		struct bt_mesh_node *node = &bt_mesh.nodes[i];
+
+		if (addr != BT_MESH_ADDR_UNASSIGNED &&
+		    node->addr == addr) {
```
Should it even be valid to call this function with such an address? Seems like a bug to me (in which case there should not be any check, or at most an assert)

trond-snekvik replied:
Isn't this the "please allocate any address" mechanism we discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/17729#discussion_r319894453)?

jhedberg replied:
Maybe I'm not understanding something, but this is a "find" function, just used for the lookups of existing provisioned nodes. Passing UNASSIGNED with the current logic is the same as directly selecting bt_mesh.nodes[0], which doesn't seem very useful?

tsvehagen replied:
@jhedberg You are right, it should never be called with the UNASSIGNED address. I'll remove the check.

@trond-snekvik No, this is not the "please allocate any address" thing :p That is done in `bt_mesh_node_alloc`.

trond-snekvik replied:
Aw, right. My bad!

tsvehagen commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -1233,6 +1303,97 @@ static void store_pending_keys(void)
 	}
 }
 
+#if defined(CONFIG_BT_MESH_PROVISIONER)
```
One problem is seems to be that the static `node_updates` array is using `CONFIG_BT_MESH_NODE_COUNT` which is only defined when `CONFIG_BT_MESH_PROVISIONER` is defined. Any way around this?

jhedberg replied:
You could have something like:
```
#if defined(CONFIG_BT_MESH_NODE_COUNT)
#define NODE_COUNT CONFIG_BT_MESH_NODE_COUNT
#else
#define NODE_COUNT 0
#endif
```

And then use NODE_COUNT instead of CONFIG_BT_MESH_NODE_COUNT

tsvehagen replied:
@jhedberg One way I guess would be to still have a `node_updates[1]` if `CONFIG_BT_MESH_PROVISIONER` is not set. I think this would be optimized away but it feels like sort of strange to do it like that.

carlescufi commented:
@tsvehagen could you please rebase this PR and reply to the unanswered review comments? Thanks!

tsvehagen commented:
> @tsvehagen could you please rebase this PR and reply to the unanswered review comments? Thanks!

I'm working on it, will hopefully find some more time tomorrow.

tsvehagen commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -734,6 +741,150 @@ u32_t bt_mesh_next_seq(void)
+ * free for use. When a conflict is found, next will be set to the next address
+ * available after the conflicting range and -EAGAIN will be returned.
+ */
+static int is_addr_free(u16_t addr_start, u8_t num_elem, u16_t *next)
```
Fixed

tsvehagen commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -734,6 +741,150 @@ u32_t bt_mesh_next_seq(void)
 	return seq;
 }
 
+#if defined(CONFIG_BT_MESH_PROVISIONER)
```
Added nodes.c and nodes.h but nodes.c is still only compiled when CONFIG_BT_MESH_PROVISIONER is enabled. For example `bt_mesh_node_find` wants to access `bt_mesh.nodes` which is only available when `CONFIG_BT_MESH_PROVISIONER` is set. This is similar to the comment below regarding `settings.h`

jhedberg replied:
Why is that an issue? As long as bt_mesh.nodes is only accessed from nodes.c there shouldn't be any issue, right? Externally (from other c-files) you can then use IS_ENABLED before calling the function.

tsvehagen commented on subsys/bluetooth/mesh/net.h:
```diff
@@ -385,3 +397,10 @@ static inline void send_cb_finalize(const struct bt_mesh_send_cb *cb,
+struct bt_mesh_node *bt_mesh_node_alloc(u16_t addr, u8_t num_elem,
+					u16_t net_idx);
+void bt_mesh_node_del(struct bt_mesh_node *node, bool store);
+#endif
```
Moved to nodes.h

tsvehagen commented:
@trond-snekvik I changed the `bt_mesh_app_key_get()` that you recently added. I added an `addr` argument in order to know what device key to fetch. Can you please check that `unseg_app_sdu_unpack()` still makes sense. is it possible to forward friend-messages that are encrypted with a device-key?

LingoMeng commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -1590,14 +1629,30 @@ void bt_mesh_heartbeat_send(void)
-	if (app_idx == BT_MESH_KEY_DEV) {
+	if (app_idx == BT_MESH_KEY_DEV_LOCAL ||
+	    (app_idx == BT_MESH_KEY_DEV_REMOTE &&
+	     addr == bt_mesh_primary_addr())) {
```
Is there a rule that dev key can only be used by primary element?

WilliamGFish replied:
I dont think this is linked to the primary element rather the primary node address:

```
u16_t bt_mesh_primary_addr(void)
{
	return dev_primary_addr;
}
```

LingoMeng replied:
> I dont think this is linked to the primary element rather the primary node address:
> 
> ```
> u16_t bt_mesh_primary_addr(void)
> {
> 	return dev_primary_addr;
> }
> ```

Isn't that the same concept? ref `Mesh Profile Spec 1.0.1 Section 2.3.4 Elements`

WilliamGFish replied:
Yes and no, the second paragraph gives a clue;

> The primary element is addressed using the first unicast address assigned to the node during provisioning.

LingoMeng replied:
I don't see the specification stating that dev_key can `only `be used by primary element.

trond-snekvik replied:
I think you're right @LingoMeng, there's no such rule, it's just that all the models that currently use devkey must be in the primary element. This check should use `bt_mesh_elem_find()` instead

tsvehagen replied:
Yes I think @LingoMeng is correct. This check is for the case when sending to a model which uses the receivers device key as app key and _we_ are the receiver. Right now this is the case when using the `cfg_cli` model for example.

I will change to `bt_mesh_elem_find()` instead.

tsvehagen replied:
Fixed

trond-snekvik commented on subsys/bluetooth/mesh/nodes.c:
```diff
@@ -0,0 +1,160 @@
+	for (i = 0; i < ARRAY_SIZE(bt_mesh.nodes); i++) {
+		struct bt_mesh_node *node = &bt_mesh.nodes[i];
+
+		if (node->addr == addr) {
```
Related to @LingoMeng's comment on primary address: This should be a range check for `node->addr` to `node->addr + node->num_elem - 1`

tsvehagen replied:
Fixed

trond-snekvik commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -1400,6 +1407,19 @@ void bt_mesh_net_start(void)
+		node = &bt_mesh.nodes[i];
+		if (node->net_idx != BT_MESH_KEY_UNUSED) {
+			bt_mesh_prov_node_added(node->net_idx, node->addr,
+						   node->num_elem);
```
I'm not sure I agree with this addition. I'd expect the provisioner app to use this callback as a trigger to start configuration of the node, something you'd only really want to do once. Wouldn't it make more sense for the app to do this on its own in init if it needs to? Currently, there's no trivial way to tell whether this callback is coming for the first time, or if it's just a repeat of a previously added node.

tsvehagen replied:
Uhm yea I think I just added it because that's how the `complete()` callback for the local node is done. I will remove it.

Currently there is no way for an application to actually get a hold of the provisioned nodes either but I was thinking that I would add that in another PR.

tsvehagen replied:
Fixed

trond-snekvik reviewed (changes requested):
> @trond-snekvik I changed the bt_mesh_app_key_get() that you recently added. I added an addr argument in order to know what device key to fetch. Can you please check that unseg_app_sdu_unpack() still makes sense. is it possible to forward friend-messages that are encrypted with a device-key?

That'll work nicely, it's only called on packets from the friend to the LPN :slightly_smiling_face:

trond-snekvik commented:
@tsvehagen What's left to do in this PR?
To me, the node_added call on start is the only thing I think has to be addressed before we can merge, the rest is mostly nitpicking and stuff that can be postponed.

tsvehagen commented:
> @tsvehagen What's left to do in this PR?
> To me, the node_added call on start is the only thing I think has to be addressed before we can merge, the rest is mostly nitpicking and stuff that can be postponed.

Yea I'm quite happy with it and I think it is ready for merge.

jhedberg commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -88,6 +88,26 @@ int bt_mesh_provision(const u8_t net_key[16], u16_t net_idx,
+
+	if (bt_mesh_subnet_get(net_idx) == NULL) {
+		return -EINVAL;
+	}
```
One of the things that has been stressed at many Mesh WG meetings I've attended is that the Provisioner is not required to be a node on the network. It's only if we also support configuration client that we need to be a node. The above two checks seem to require that we're both provisioned into a network and that we've been configured with the net_idx we intend to give to the to-be-provisioned node.

What you should be checking here (I think) is if the net_idx is available in the provisioner's database.

While we could just say that we don't support separating configuration client and provisioner roles, we should still have separation of the provisioner's database and the parts of it that have been configured for the local node. We could e.g. be adding app & net keys to other nodes that the local node (configuration client) hasn't configured for its own use as a node (beyond passing values as parameters in configuration client messages).

tsvehagen replied:
I think it would be a very good idea to have that sort of database that could have its own API for adding keys and such. The nodes added should probably also be put in that database instead of in the `struct bt_mesh_net`.

> What you should be checking here (I think) is if the net_idx is available in the provisioner's database.

Agree, but there is no notion of a provisioner's database currently. Do you think that is required before merging this PR?

The checks in the API call is basically just a "fail fast" but I'm pretty sure they could be removed. The real problem is that `send_prov_data()` uses the local node as database for the `net_key`, `iv_index` and `flags`. I think this is where the separation should happen.

For know I think it would be okay to say that separation is not supported but I see no problem of creating the separation in the future.

jhedberg replied:
@tsvehagen as long as there's no major impact on the public API I think we can do the separation (provisioner database) in a later PR, but it'd be good to be done sooner than later, ideally before the 2.1 merge window closes in a couple of weeks.

tsvehagen replied:
Okay so regarding the API I think we should then have a short discussion before merging.

Currently the API only takes a `net_idx`. This won't be enough if for example a provisioner can work with multiple networks as a the same `net_idx` could appear in both. In that case there must also be some sort of network id parameter. Is that something we want to support?

tsvehagen replied:
> One of the things that has been stressed at many Mesh WG meetings I've attended is that the Provisioner is not required to be a node on the network. It's only if we also support configuration client that we need to be a node. The above two checks seem to require that we're both provisioned into a network and that we've been configured with the net_idx we intend to give to the to-be-provisioned node.

@jhedberg 
One thing that I don't really understand is that during provisioning, like I commented above, the provisioning data includes `flags` and `iv_index`. If the provisioner device is separate from the configuration client node, how will the provisioner know these things? Is there anything specified for how a provisioner can get new network information?

jhedberg replied:
>  If the provisioner device is separate from the configuration client node, how will the provisioner know these things? Is there anything specified for how a provisioner can get new network information?

This is left as a vendor-specific solution for the provisioner implementation. There's a spec coming out soon with a standard format for encoding such information, which can be used to synchronise multiple provisioners of the same network (this e.g. lets provisioners get subsets of the unicast address range so they don't go and provision different devices with the same addresses).

> Also in the Mesh Profile spec under 3.10.1 Mesh Network Creation procedure, it says that a Provisioner shall allocate a unicast address and assign it to its primary element. Doesn't this make it a node?

That's indeed quite odd, and in contradiction to what I've learned from listening and talking to  Mesh WG members. I'll check if upcoming spec versions have this clarified/corrected.

@trond-snekvik do you have any thoughts on these issues?

trond-snekvik replied:
> @trond-snekvik do you have any thoughts on these issues?

I'm also under the impression that the WG wants provisioners to be some omnipresent non-node entity, but I think the most natural way to implement it is to combine it with the configurator role.

I'm fine with the current API and its checks. Forcing the users to provide the netkey value in addition to the net_idx just introduces awkwardness for the normal case, and the current approach isn't really a blocker for the non-node provisioner.

jhedberg reviewed (approved):
Approving, but with the expectation that the separation of provisioner database happens fairly soon (i.e. that the provisioner is not required to be a node with all keys etc that get provisioned & configured to other nodes)

jhedberg commented:
Once @trond-snekvik has also approved I think we can merge this.

tsvehagen commented:
> Approving, but with the expectation that the separation of provisioner database happens fairly soon (i.e. that the provisioner is not required to be a node with all keys etc that get provisioned & configured to other nodes)

For me it's not a rush to merge it. I can continue to work a a provision database and try to have that it place before moving forward with this. It might not be done before 2.1 merge window closes though.

joerchan commented:
@jhedberg Looks like you have the approval you needed. :)

WilliamGFish commented:
Just fetched these changes and the BBC Microbit is not happy. Mesh demo went from almost working to:

> ***** Booting Zephyr OS build zephyr-v2.0.0-1610-g3282ff22c4aa *****
> Initializing...
> Unicast address: 0x0003
> Bluetooth init failed (err -22)
> [00:00:00.015,136] [1B][1;31m<err> bt_settings: settings_subsys_init failed (err -22)[1B][0m

I will raise an issue
